module/zfs/ddt_zap.c

```diff
@@ -71,11 +76,15 @@ ddt_zap_decompress(uchar_t *src, void *dst, size_t s_len, size_t d_len)
 {
 	uchar_t version = *src++;
 	int cpfunc = version & DDT_ZAP_COMPRESS_FUNCTION_MASK;
-	zio_compress_info_t *ci = &zio_compress_table[cpfunc];
-
-	if (ci->ci_decompress != NULL)
-		(void) ci->ci_decompress(src, dst, s_len, d_len, ci->ci_level);
-	else
+	int err;
+
+	abd_t sabd, dabd;
+	abd_get_from_buf_struct(&sabd, src, s_len);
+	abd_get_from_buf_struct(&dabd, dst, d_len);
+	err = zio_decompress_data(cpfunc, &sabd, &dabd, s_len, d_len, NULL);
+	abd_free(&dabd);
+	abd_free(&sabd);
+	if (err != 0)
 		memcpy(dst, src, d_len);
```

Comment:
I am not sure it is an issue, but it seems a change of the logic.  Previously `memcpy()` was called only for "algorithms" without a decompressors, while decompression errors were ignored, for good or bad.  Now though `memcpy()` is called also in case of errors, returning compressed data instead of possibly partly decompressed.

Reply:
Ahh yeah, that's true. It wasn't really possible for it to happen because the compression type is hardcoded anyway, but I've put the NULL test back in anyway, and made the compress call `VERIFY0()`, in keeping with the style of the rest of the dedup code.

I'm content with that for now; unamazing but there's a lot more to be done with dedup before the end of time.

Reply:
I was thinking towards reporting error rather than asserting it, may be it could be less fatal, but whatever.

Reply:
Yeah, you're not wrong, but I didn't want to get into changing dedup again this decade :sweat_smile: Thanks!